[mlir][Vector] Add narrow type emulation pattern for vector.maskedload

tyb0807 opened the pull request.

tyb0807 committed:
[mlir][Vector] Add narrow type emulation pattern for vector.maskedload

As for vector.load and vector.transfer_read ops, vector.maskedload op
with narrow bitwidth (e.g., i4) can now be converted to supported wider
bitwidth (e.g., i8 or i32).

tyb0807 committed:
[mlir][Vector] Handle narrow type emulation of vector.maskedload when mask is an extraction

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -103,6 +104,212 @@ struct ConvertVectorLoad final : OpConversionPattern<vector::LoadOp> {
+    //
+    // Since we are effectively loading 16 bits (2xi8) from the memref with the
+    // new mask, while originally we only wanted to effectively load 12 bits
+    // (3xi4) from the memref, we need to set the second half of the last i8
+    // that was effectively loaded (i.e. the second i8) to 0.
```
Trying to understand if this is really needed. Given that we are only emulating the vector loads/stores and bitcasting them back to the i4 type (see vector load case), it should be ok to just do an `arith.select` using the original mask, the loaded i4 vector and the original passthru value, right? If that is the case, all the code below would be really simplified!

dcaballe replied:
I'm probably the one missing something :) I meant:

```
%i8ld = vector.maskedload %0[%linear_index], %new_mask, %new_pass_thru ...
%i4ld = vector.bitcast %i8ld : vector<3xi8> to vector<6xi4>
%final_ld = arith.select %mask, %i4ld, %pass_thru : vector<6xi1>, vector<6xi4>
```

Would this work? The generated asm may not be ideal but it should definitely reduce the complexity of this lowering.

dcaballe reviewed (commented):
Awesome! Just a comment for now about the passthru value.

Would you also mind adding some description to the commit message? That would help. Thanks!

tyb0807 committed:
[mlir][Vector] Use a simpler lowering when emulating narrow type for vector.maskedload

arith.select should be used instead of a series of manual mask
manipulating ops (arith.and/or/extsi)

tyb0807 committed:
[mlir][Vector] Add integration test for vector.maskedload narrow type emulation pattern

dcaballe reviewed (approved):
LG! If you can remove the commented code that should be it! Thanks a lot for looking into this!

tyb0807 commented:
Thanks for the review!

hanhanW commented on mlir/test/Integration/Dialect/Vector/CPU/test-rewrite-narrow-types.mlir:
```diff
@@ -164,6 +164,13 @@ func.func @fext(%a: vector<5xi8>) {
   return
 }
 
+func.func @fcst_maskedload(%A: memref<?xi4>, %passthru: vector<6xi4>) -> vector<6xi4> {
+  %c0 = arith.constant 0: index
+  %mask = vector.constant_mask [3] : vector<6xi1>
+  %1 = vector.maskedload %A[%c0], %mask, %passthru :
+    memref<?xi4>, vector<6xi1>, vector<6xi4> into vector<6xi4>
+  return %1 : vector<6xi4>
+}
```
I just noticed that this is not tested because the patterns are not populated..

https://github.com/llvm/llvm-project/blob/a0bd6361d4840f159e48c7359807ae279d006fb0/mlir/lib/Dialect/Vector/TransformOps/VectorTransformOps.cpp#L162-L165

To test it, we need to set up a type converter and populate those patterns.
